File: src/main/java/de/ng/ahnenwin2gedcom/gedcom/DateFormatter.java
package de.ng.ahnenwin2gedcom.gedcom;

import de.ng.ahnenwin2gedcom.helperfunctions.StringFun;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeFormatterBuilder;
import java.time.format.DateTimeParseException;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.function.Predicate;

class DateFormatter {
    private static final DateTimeFormatter dateTimeFormatter = new DateTimeFormatterBuilder()
            .parseCaseInsensitive()
            .appendPattern("dd MMM yyyy")
            .toFormatter()
            .withLocale(Locale.ENGLISH);

    private static final Predicate<Integer> isInMonthRange = month -> 0 < month && month < 13;

    static String format(String day, String month, String year) {
        String date = null;
        String formattedYear = formatYear(year);
        String formattedMonth = formatMonth(month);
        String formattedDay = formatDay(day);
        if (formattedYear != null) {
            date = formattedYear;
            if (formattedMonth != null) {
                date = formattedMonth + " " + date;
                if (formattedDay != null && canParseDate(formattedDay, formattedMonth, formattedYear)) {
                    date = formattedDay + " " + date;
                }
            }
        }
        return date;
    }

    static boolean hasInvalidStructure(String day, String month, String year) {
        String formattedDay = DateFormatter.formatDay(day);
        String formattedMonth = DateFormatter.formatMonth(month);
        String formattedYear = DateFormatter.formatYear(year);
        boolean dayDefinedAndInvalid = StringFun.notEmpty(day) && formattedDay == null;
        boolean monthDefinedAndInvalid = StringFun.notEmpty(month) && formattedMonth == null;
        boolean yearDefinedAndInvalid = StringFun.notEmpty(year) && formattedYear == null;
        boolean dayDefinedButNotYearOrMonth = StringFun.notEmpty(day) && (formattedMonth == null || formattedYear == null);
        boolean monthDefinedButNotYear = StringFun.notEmpty(month) && formattedYear == null;
        return dayDefinedAndInvalid
                || monthDefinedAndInvalid
                || yearDefinedAndInvalid
                || dayDefinedButNotYearOrMonth
                || monthDefinedButNotYear;
    }

    private static boolean canParseDate(String day, String month, String year) {
        String dateString = String.format("%s %s %s", day, month, year);
        try {
            LocalDate.parse(dateString, dateTimeFormatter);
        } catch (DateTimeParseException ignore) {
            return false;
        }
        return true;
    }

    static String formatDay(String day) {
        return parseIntegerValue(day)
                .map(intDay -> String.format("%02d", intDay))
                .orElse(null);
    }

    static String formatMonth(String monthNumber) {
        return parseIntegerValue(monthNumber)
                .filter(isInMonthRange)
                .map(intMonth -> DateFormat.MONTH_ABBREVIATIONS[intMonth - 1])
                .orElse(null);
    }

    static String formatYear(String year) {
        return parseIntegerValue(year)
                .map(intYear -> String.format("%04d", intYear))
                .orElse(null);
    }

    static String formatAsNote(String notePrefix, String day, String month, String year) {
        List<String> notes = new LinkedList<>();
        if (StringFun.notEmpty(day)) notes.add("Tag: " + day);
        if (StringFun.notEmpty(month)) notes.add("Monat: " + month);
        if (StringFun.notEmpty(year)) notes.add("Jahr: " + year);
        return notePrefix + ": " + String.join(", ", notes);
    }

    private static Optional<Integer> parseIntegerValue(String value) {
        if (StringFun.isEmpty(value)) return Optional.empty();
        Integer intValue;
        try {
            intValue = Integer.parseInt(value);
        }
        catch (NumberFormatException ignore) {
            intValue = null;
        }
        return Optional.ofNullable(intValue);
    }
}

File: src/main/kotlin/de/ng/ahnenwin2gedcom/gedcom/DateFormatter.kt
package de.ng.ahnenwin2gedcom.gedcom

import de.ng.ahnenwin2gedcom.helperfunctions.StringFun.notEmpty
import java.time.LocalDate
import java.time.format.DateTimeFormatterBuilder
import java.time.format.DateTimeParseException
import java.util.*

internal object DateFormatter {

    private val dateTimeFormatter = DateTimeFormatterBuilder()
        .parseCaseInsensitive()
        .appendPattern("d MMM y")
        .toFormatter()
        .withLocale(Locale.ENGLISH)

    @JvmStatic
    fun format(day: String?, month: String?, year: String?): String? {
        var date: String? = null
        val formattedYear = year?.let { formatYear(it) }
        val formattedMonth = month?.let { formatMonth(it) }
        val formattedDay = day?.let { formatDay(it) }
        if (formattedYear != null) {
            date = formattedYear
            if (formattedMonth != null) {
                date = "$formattedMonth $date"
                if (formattedDay != null && canParseDate(formattedDay, formattedMonth, formattedYear)) {
                    date = "$formattedDay $date"
                }
            }
        }
        return date
    }

    @JvmStatic
    fun hasInvalidStructure(day: String?, month: String?, year: String?): Boolean {
        val formattedDay = day?.let { formatDay(it) }
        val formattedMonth = month?.let { formatMonth(it) }
        val formattedYear = year?.let { formatYear(it) }
        val dayDefinedAndInvalid = notEmpty(day) && formattedDay == null
        val monthDefinedAndInvalid = notEmpty(month) && formattedMonth == null
        val yearDefinedAndInvalid = notEmpty(year) && formattedYear == null
        val dayDefinedButNotYearOrMonth = notEmpty(day) && (formattedMonth == null || formattedYear == null)
        val monthDefinedButNotYear = notEmpty(month) && formattedYear == null
        return (dayDefinedAndInvalid
                || monthDefinedAndInvalid
                || yearDefinedAndInvalid
                || dayDefinedButNotYearOrMonth
                || monthDefinedButNotYear)
    }

    private fun canParseDate(day: String, month: String, year: String): Boolean {
        val dateString = "$day $month $year"
        try {
            LocalDate.parse(dateString, dateTimeFormatter)
        } catch (ignore: DateTimeParseException) {
            return false
        }
        return true
    }

    private fun formatDay(day: String): String? {
        return parseIntegerValue(day)?.toString()
    }

    private fun formatMonth(monthNumber: String): String? {
        val monthInt = parseIntegerValue(monthNumber).takeIf { it in 1..12 } ?: return null
        return DateFormat.MONTH_ABBREVIATIONS[monthInt - 1]
    }

    private fun formatYear(year: String): String? {
        return parseIntegerValue(year)?.toString()
    }

    @JvmStatic
    fun formatAsNote(notePrefix: String, day: String?, month: String?, year: String?): String {
        val notes: MutableList<String> = LinkedList()
        if (notEmpty(day)) notes.add("Tag: $day")
        if (notEmpty(month)) notes.add("Monat: $month")
        if (notEmpty(year)) notes.add("Jahr: $year")
        return notePrefix + ": " + java.lang.String.join(", ", notes)
    }

    private fun parseIntegerValue(value: String): Int? {
        return try {
            value.toInt()
        } catch (ignore: NumberFormatException) {
            null
        }
    }
}
